Reduce paragraph block CSS specificity

m-e-h opened the pull request:
## Description
Move drop-cap:focus style to editor.css.
Reduce specificity on front-end drop-cap by removing the `:not(:focus)` pseudo-class.
Show a visual hint that a dropcap is present while focused by:

- make the first letter a little larger, `1.2em`
- make the first letter bolder by using a text-shadow of the same color
- keep the uppercase text-transform. 


## How has this been tested?
Visual tests in browser.

## Screenshots <!-- if applicable -->
![dropcap](https://user-images.githubusercontent.com/3585901/50236414-55b6f500-0388-11e9-865f-80cdc81c7a02.gif)


## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

m-e-h committed:
move dropcap focus to editor.css. Show hint of DC on focus

m-e-h committed:
Move dropcap focus to editor.css. Reduce specificity

m-e-h commented:
This PR also helps to chip away at some of
https://github.com/WordPress/gutenberg/issues/11677
and some of
https://github.com/WordPress/gutenberg/issues/11668

jorgefilipecosta reviewed (commented):
Hi @m-e-h, I agree with the logic of these changes, I think focus styles that are editor specific should not be referenced on the front end styles.
Unfortunately, I think we will not be able to differentiate the first letter with this styles, as even this simple styles change causes the bug shown in the following gif:
![dec-20-2018 15-30-34](https://user-images.githubusercontent.com/11271197/50293922-9b4cee00-046c-11e9-8a3e-aec83b36ab04.gif)

m-e-h commented:
Thanks for looking at this @jorgefilipecosta!

That appears to be a Chrome bug. Works fine in Firefox. 
I'll see if I can find out what's going on.

m-e-h commented:
Yep, it's a known bug in Chrome. https://bugs.chromium.org/p/chromium/issues/detail?id=174349
Probably accounts for some of the earlier issues the editor had with Drop cap. 

The cursor bug becomes present even with an empty rule in the stylesheet:
```
.has-drop-cap:focus::first-letter {}
```
So it looks like `:not(:focus)` is the only solution. At least until this gets fixed in Chrome.

m-e-h commented:
Though it would be nice to ultimately remove that `:not(:focus)` style from the front end, I be don't see any way around it in the foreseeable future. Not much activity on that Chromium bug page. 

Should I close this? 

One sort of separate thing I'd also like to do here with the dropcap styles is remove the `p` qualifier preceding each selector.

Does anyone know if there's a reason for it being there?

jasmussen commented:
Nice work here. Some GIFs will help.

This branch:

![this branch](https://user-images.githubusercontent.com/1204802/52469656-5b60d800-2b8c-11e9-827c-2b65968481a6.gif)

Master:

![master](https://user-images.githubusercontent.com/1204802/52469658-5dc33200-2b8c-11e9-9c5c-aaafed55f551.gif)

I appreciate what you're trying to do here, *notably* the reduction in CSS specificity. But the bold first letter goes a little bit against the principle that the unselected block is a preview and the selected block can be an edit mode. I deeply appreciate what you're trying to do here, but honestly I feel like what we have in master with regards to the select and edit behavior is the better solution.

But the change to CSS specificity in style.scss is a total no-brainer. 

Can you change this PR so the only change is the reduction in CSS specificity in the style.scss file? I think that'd be worth shipping.

m-e-h committed:
Merge remote-tracking branch 'upstream/master' into fix/dropcap-visual-on-focus

m-e-h committed:
Easing up on paragraph block specificity.

m-e-h commented:
I removed the dropcap changes that I tried on the editor.scss.
I removed the qualifying `p` tag from each selector in style.scss, thereby decreasing the specificity. 

But @jasmussen I'm bringing back the editor dropcap argument if and when Chrome ever fixes it's `:focus:first-letter` bug. :smiley:

kjellr reviewed (approved):
Looks good, thanks @m-e-h!

ZebulanStanphill commented on packages/block-library/src/paragraph/style.scss:
```diff
@@ -1,46 +1,44 @@
-	&.is-small-text {
-		font-size: 14px;
-	}
+.is-small-text {
```
Isn't this selector a bit _too_ generic? What about other blocks that use a class with this same name? Shouldn't this class be either namespaced (e.g. `.wp-block-paragraph__is-small-text` or changed to something along the lines of `.wp-block-paragraph.is-small-text`?

m-e-h replied:
Probably should've been `.wp-block-paragraph__is-small-text`. 
In the case of this PR though, does the `p` qualifier really make it any less generic? 

I think `.wp-block-paragraph.is-small-text` is too specific for a core style. Especially with something like `font-size` which is likely one of the first things that a theme will override.

m-e-h replied:
For the record though, I would prefer `.wp-block-paragraph.is-small-text` over `p.is-small-text`. Since there are more appropriate hooks in place for themes to override the Paragraph Block font-sizes.

As for being generic, the one that sticks out to me here is `.has-background`. That's not even "text" specific and we're adding some hefty padding here. Is that class used in any other blocks?

m-e-h replied:
Dang! This is going to have to be fixed. The button also uses `.has-background` and is getting the padding from the Paragraph block CSS. Guess We're going to have to increase specificity for the paragraph block rather than decrease. 

@kjellr How should I go about fixing? A new PR?

kjellr replied:
Ah yes, good catch. I missed that selector. Open up a new PR and I'll take a look shortly.

kjellr replied:
I think the `has-background` and `has-text-color` selectors are the only ones that definitely should be changed for now. I don't believe the text size and dropcap ones are used elsewhere.

kjellr replied:
Had a few minutes, so I opened #13821 to take care of this. 👍

m-e-h replied:
Thanks @kjellr! 
Totally my fault. The PR was originally for the Dropcap and sort of morphed into being about specificity. I didn't put as much thought into it as I should have.

I agree that the other selectors don't look as problematic as `has-background`. Even if something like `is-small-text` is used in other blocks, we're only setting the font size, which would be expected for such a class name and would likely be the same or overridden by the other block.

It's the `padding` on a generic "background" selector that made this troublesome. 
It could be debated that the padding style is more appropriate in `theme.css`. But that's a topic for another day. :smile:

jasmussen replied:
You're all amazing. Thanks for the hard work here.

kjellr replied:
> Totally my fault. The PR was originally for the Dropcap and sort of morphed into being about specificity. I didn't put as much thought into it as I should have.

No problem at all! It's all set now. Thanks again for your contribution, and thanks @ZebulanStanphill for starting the discussion that led to us catching this. 👍

youknowriad commented on packages/block-library/src/paragraph/style.scss:
```diff
@@ -1,46 +1,44 @@
-	}
+.is-larger-text {
+	font-size: 48px;
+}
```
Is this the right file for these stylesheets? If I remember properly, these are theme options + editor styles? shouldn't we move these styles to the build-int `editor-styles.css` instead?
